src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
+
+            foreach (var documentId in args.Select(a => a.DocumentId).Distinct())
+            {
+                analyzer.RemoveDocument(documentId);
```

Comment:
We have very differnt views on what is "clear" or not :)

When i look at this code, it is not at all clear.  When i see this code, i see a lot of brittleness and fragility.  Which, IMO, is part of the reason we had this leak in the first place.

1. There was no way to tell from looking at the original code that there was even a problem.  
2. All we started with was an "IIncrementalAnalyzerProvider".  There's no clear indication that "it will never get that analyzer".
3. The code is very confusing because it says it "Creates" analyzers, but is actually retrieving cached analyzers.
4. The code is confusing because it depends on knowing precisely how one specific analyzer works and what is necessary for it to clear it's state.
5. It's not clear at all how we know we're even getting the right set of diagnostics to clear.  In 'Register' we use the set of open documents, but in Unregister we use an entirely different set of documents.
6. Just as there was no way to tell originall that there was a problem, there's practically no way for me to tell right now that this is a sufficient solution.  

In other words, this code may work, but only because it was written with every part of the system having deep an intimate understanding of the internal workings of other parts of the system.  That is a brittle solution.  The moment we change *anything* in that analyzer, we may break this and we may have a leak again.

Furthermore, as i've pointed out above and in this discussion, the very fact that 10 lines of code spawned 70+ comments back and forth indicates precisely how confusing and subtle this is, and how difficult it will be to maintain.  How would anyone else know what the right thing to do would be if they needed to come here and fix up this code 3, 6, 12 months down the line?  

The code doesn't cleanly separate out concerns, and it strongly couples internal implementation details of disparate components.  As a one-off fix to quickly solve a memory concern, i may sign off on it.  But i would be really against this code flowing into our main branch, and i think we should come up with some better ways of fixing this issue in a much cleaner and more self describing manner.